### ISSUE TITLE
Provide argument resolver for KeysetScrollPosition

### ISSUE THREAD

**Issue by vpavic** [spring-projects/spring-data-commons#3148]
In order to make `KeysetScrollPosition` easier to use for developers, especially having in mind its non-trivial nature (in contrast to `OffsetScrollPosition`), Spring Data Commons should provide the appropriate argument resolver.

Typically something like `KeysetScrollPosition` would be serialized to JSON and then base64 encoded.

**Comment by quaff** [comment]
My proposal: https://github.com/spring-projects/spring-data-commons/issues/2856#issuecomment-1596380892

>> The key problem is how to convert keys to correct type, maybe we should introduce an annotation to extract actual type from underlying domain object, for example
>>
>> ```@TypeHints(from = User.class) KeysetScrollPosition position```

**Comment by mp911de** [comment]
Generally speaking, I would like to see some folks implementing serialization and deserialization for these types. As we're working with message structures that could potentially reveal application internals, integrity and encryption are likely to be part of such a story. For the time being, we do not want to include such a change.

**Comment by quaff** [comment]
> As we're working with message structures that could potentially reveal application internals

Could you elaborate more? as far as I know, key names is not sensitive information. 

> integrity and encryption are likely to be part of such a story.

I think it's not restful, we should allow client construct query by specifying parameters.

**Comment by mp911de** [comment]
We can always reopen closed tickets. It would be great to have a bit of sample code to get a different perspective.

**Comment by mp911de** [comment]
> Could you elaborate more? as far as I know, key names is not sensitive information.

Indeed, key (property) names are not sensitive. However, the key values might be sensitive and not exposed. We require keys and values to reconstruct the query.

> I think it's not restful, we should allow client construct query by specifying parameters.

No one said it has to be restful. Think about `ScrollPosition` as cursor that is generated on the server side. A good comparison is Cassandra's `PagingState` that looks to users like a bunch of random bytes to continue a scrolling session. As external client, you cannot expect applications to present a scroll position that the application has never generated in the first place.

**Comment by mceste-sapns2** [comment]
There seems to be 2 challenges with implementing this: 1) how keysets are encoded, and 2) where the translation happens.

We’ve implemented our own `KeysetPageable` (many years before `ScrollPosition`existed) and we encode/decode in the repository using a ConversionService since it’s aware of the sort order and the property types. Additionally, clients don’t need to construct keysets or really care how they’re encoded, they just need to pass them back to continue iteration. i’m happy to share a code sample if it would help. We haven’t switched to KeysetScrollPosition yet because we’d need to implement our own argument resolver, which would be tricky given the lack of type information available as a controller method parameter. We pushed the parsing into the Repository, which means we wouldn’t be able to leverage the out of the box Spring Data JPA repository.
